### PR TITLE
Update and refactor

### PR DESCRIPTION
**Pull request by jmarshrossney**
As useful as this repo and scivision have been for getting started, we are increasingly running into issues which would be much easier to resolve if we had our own 'version' of this model to build on and keep up to date.

As a first step, I've made minimal changes so that it runs without complaining using Python 3.11 and PyTorch 2.4, switched to the `pyproject.toml` standard for packaging and replaced `requirements.txt` with `requirements.in`.

### CODE REVIEW

**Comment by jmarshrossney** [comment]
Just a note to self about the labelling in the `utils` module. There are 5 different 'experiments' which correspond to a different number of output classes, from `label1` which is just plankton vs detritus, up to `label3_detritus` which has 38 plankton classes + detritus.

The important place these labels are used is in the model `__init__`, where the final `Linear` layer (called `fc`) of ResNet is replaced by another `Linear` layer with an output size that is the number of classes for which predictions are needed.

However, afaik the only weights that have been published are [here](https://zenodo.org/records/6143685) and correspond to `label2_detritus`.

So, if we are starting from scratch we remove all of the logic around the labels, though keeping these experiments in mind for future reference.

**Comment by jmarshrossney** [comment]
A few pieces of information are missing, which were probably in the original train and test scripts that didn't get published (afaik), though also might be recoverable from the [Data Study Group report](https://zenodo.org/records/6799166).

### Dataset

The `ds` argument to `PlanktonDataset` is an `xarray.Dataset` which _appears_ to contain a set of plankton images that have been padded such that they are the same size and can be concatenated. I think this because it looks like in `__getitem__` a single image is selected and the padding is discarded:

```python
im_raw = self.ds.sel(concat_dim=self.img_ixs[idx])
imw = im_raw.image_width.values
iml = im_raw.image_length.values

im_ori = im_raw['raster'][0:iml, 0:imw, :].values
```

It also seems that `ds` carries some named dimensions (e.g. `concat_dim`) but it's probably reasonable to assume there's nothing more going on here than selecting dimensions by label, just confused me for a bit since I didn't realise the named dimensions could be accessed using `.name` rather than `["name"]`!

Interestingly, after throwing away the padding to get the original image, that image is then resized to `[256, 256]` using the default (bilinear) interpolation. I have two questions:

1. This seems like it might cause problems since the classifier now effectively has kind of marginalise over different resolutions..?
1. If one is going to resize all images to 256x256, wouldn't it be sensible to do this once for each image and create a new dataset for these 256x256 images, rather than doing it every time you want to load an image? I suppose at that point you may as well just save the dataset as a massive tensor. Maybe I'm missing something here, e.g. these conversions are a small overhead compared to the cost of the forward + backward pass.

### Training

What was the loss?

For prediction there is currently just `_, preds = torch.max(y, 1)` but presumably the outputs were softmaxed during training. Would be nice to be sure.

**Comment by jmarshrossney** [comment]
Since I have broken it anyway, I am removing the `.scivision` directory which contains the configuration that allows this model to be loaded through scivision.

**This is something I want to revisit in future**, ideally in a world in which the rewrite of [`intake`](https://github.com/intake/intake) has been completed and scivision have updated their spec.

**Comment by jmarshrossney** [comment]
There is not much to really do with the benchmark / test set of 26 images, except check that the results are reasonably consistent with a model that's been well-trained. To this end I added a couple of metrics to the benchmark script.

I used `torchmetrics` because I agree with their principle of standardising metrics across ML, but unless this repo ends up using any more metrics I will probably remove in future in favour of pure-torch implementations of accuracy and confusion matrix.

**Comment by jmarshrossney** [comment]
Thought about making `torchmetrics` and `matplotlib` optional dependencies, but it's just overengineered. There is no need to use either of them for the benchmark as it currently stands so just dropping them for now.

**Comment by jmarshrossney** [comment]
The current version has been tested with a fresh 3.11.9 `venv` and a `python -m pip install .` followed by running the benchmark `cefas-benchmark`.

**Comment by jmarshrossney** [comment]
I have checked that pip installing from the repository (`@update`) and running the benchmark works..
